core/src/miner/backup.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use ckey::Public;
+use kvdb::{DBTransaction, KeyValueDB};
+use primitives::H256;
+use rlp::Encodable;
+
+use super::mem_pool_types::{MemPoolItem, PoolingInstant};
+use crate::db as dblib;
+
+pub struct RecoveredData {
+    pub by_hash: HashMap<H256, MemPoolItem>,
+    pub first_seqs: HashMap<Public, u64>,
+    pub last_time: PoolingInstant,
+    pub last_timestamp: u64,
+    pub next_transaction_id: u64,
+}
+
+pub fn backup_item(db: &KeyValueDB, key: H256, item: &MemPoolItem) {
```

Comment:
How about making the `backup_item` receives `mut DBTransaction` and just call `batch.put`?
This function always called in the `for` statement. It is better to collect the DB transactions and write once, not every time.

Reply:
Thank you, now `backup_~` functions receives mutable DBTransaction reference and `db.write` is called when `add_items` or `remove` function ends